Client/SafeBrowsing/GSafeBrowsingCache.swift

```diff
@@ -0,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+private struct ThreatCache: Hashable {
+    var time: TimeInterval
+    let hash: String
+    let threatType: ThreatType
+}
+
+enum SBCacheHitType {
+    case miss
+    case positive
+    case negative
+}
+
+struct SBCacheResult {
+    let threats: [ThreatType]
+    let cacheResult: SBCacheHitType
+}
+
+class SafeBrowsingCache {
+    private let cacheLock = NSRecursiveLock()
+    private var positiveCache = [String: [ThreatCache]]()
+    private var negativeCache = [String: TimeInterval]()
+    
+    func find(_ hash: String) -> SBCacheResult {
+        cacheLock.lock(); defer { cacheLock.unlock() }
+        
+        if let threats = positiveCache[hash] {
+            var results = [ThreatType]()
+            
+            for threat in threats {
+                if Date().timeIntervalSince1970 > threat.time  {
+                    results.append(threat.threatType)
+                } else {
+                    return SBCacheResult(threats: [], cacheResult: .miss)
```

Comment:
I'm a little confused with the flow here. It looks like we go through the cached entries and then if we find a single expired one we return a cache miss? I think we should instead keep looking at the other entries for that hash, it's possible there will be a valid one.

Reply:
No we shouldn't. Implementation is correct. If there is anything wrong at all in the cache and there's an expired `HashPrefix` for any `FullHash` we automatically ask the server to double check if the hash really is safe.

Reply:
You're right, that's exactly how it's [specified](https://developers.google.com/safe-browsing/v4/caching#consulting-the-cache).